cpp/convert_json_to_marc.cc

```diff
@@ -414,20 +443,77 @@ std::string ReferencedJSONDataStateToString(const ReferencedJSONDataState refere
 }
 
 
+// Expects references of the form "map:map_name{json_path}map_key" followed by an optional ",first".
+bool SplitMapReference(std::string reference, std::string * const map_name,
+                       std::string * const json_path, std::string * const map_key, bool * const first)
+{
+    // It is imperative that *first will be set, even if we bail out early!
+    *first = StringUtil::EndsWith(reference, ",first");
+
+    if (not StringUtil::StartsWith(reference, "map:"))
+        return false;
+
+    map_name->clear();
+    json_path->clear();
+    map_key->clear();
+
+    if (*first) // Strip off the trailing ",first".
+        reference.resize(reference.size() - __builtin_strlen(",first"));
+
+    auto ch(reference.cbegin() + __builtin_strlen("map:"));
+    while (ch != reference.cend() and *ch != '{')
+        *map_name += *ch++;
+    if (ch == reference.cend())
+        return false;
+    ++ch; // Skip over '{'.
+
+    while (ch != reference.cend() and *ch != '}')
+        *json_path += *ch++;
+    if (ch == reference.cend())
+        return false;
+    ++ch; // Skip over '}'.
+
+    if (*ch != ':')
+        return false;
+    ++ch; // Skip over ':'.
+
+    while (ch != reference.cend())
+        *map_key += *ch++;
+
+    return not map_name->empty() and not json_path->empty() and not map_key->empty();
+}
+
+
 ReferencedJSONDataState CategorizeJSONReferences(const std::shared_ptr<const JSON::ObjectNode> &object,
-                                                 const std::vector<std::pair<char, std::string>> &subfield_codes_to_json_tags,
+                                                 const std::vector<std::pair<char, std::string>> &subfield_codes_to_json_paths,
                                                  size_t * const common_array_length)
 {
     unsigned array_references_count(0), subfield_data_found_count(0);
     size_t last_array_length(std::numeric_limits<size_t>::max());
-    for (const auto &subfield_code_and_json_tag : subfield_codes_to_json_tags) {
-        const auto node(object->deepResolveNode(subfield_code_and_json_tag.second));
+    for (const auto &subfield_code_and_json_path : subfield_codes_to_json_paths) {
+        std::string relative_json_path(subfield_code_and_json_path.second);
+
+        // Two types of relative JSON path references exist: 1) the direct one and
+        // 2) one nested within a "map:" expression.
+        std::string map_name, json_path, map_key;
+        bool use_first_array_element_only;
+        if (SplitMapReference(subfield_code_and_json_path.second, &map_name, &json_path, &map_key,
```

Comment:
Shouldn't map_name and map_key be used somehow?

Reply:
Not here.  This is just for a preliminary check of the type of JSON references.  At other call sites of SplitMapReference, map_name and map_key are indeed used.